extensions/amp-image-slider/0.0/amp-image-slider.js

```diff
@@ -0,0 +1,566 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Animation} from '../../../src/animation';
+import {CSS} from '../../../build/amp-image-slider-0.0.css';
+import {bezierCurve} from '../../../src/curve';
+import {dev, user} from '../../../src/log';
+import {htmlFor} from '../../../src/static-template';
+import {isExperimentOn} from '../../../src/experiments';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {numeric} from '../../../src/transition';
+import {setStyle} from '../../../src/style';
+
+// event-helper.js -> listen; option: passive = true
+// mount both mouse & touch listener ()
+// Test: mainly describes.integration
+// Visual diffing tests
+
+// Wrap <a> anchor around amp-image
+// This is actually creating some trouble with gesture
+
+// this.mutateElement(callback), in which move children
+
+// INTERESTING EXPERIMENT:
+// getLayoutBox()?
+// this seems to be remeasured and updated on viewport update
+
+// TODO(kqian): fix gesture collision between drag mouseup and click to move
+export class AmpImageSlider extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {Document} */
+    this.doc_ = this.win.document;
+
+    /** @private {boolean} */
+    this.isMobile_ = /Android|iPhone|iPad|iPod/i.test(this.win.navigator.userAgent);
+
+    /** @private {boolean} */
+    this.isHoverSlider_ = (this.element.getAttribute('type') === 'hover')
+        && !this.isMobile_; // coerce to drag slider on mobile
+
+    /** @private {!Element} */
+    this.container_ = this.doc_.createElement('div');
+
+    /** @private {?Element} */
+    this.leftAmpImage_ = null;
+
+    /** @private {?Element} */
+    this.rightAmpImage_ = null;
+
+    /** @private {?Element} */
+    this.leftLabelWrapper_ = null;
+    /** @private {?Element} */
+    this.leftLabel_ = null;
+
+    /** @private {?Element} */
+    this.rightLabelWrapper_ = null;
+    /** @private {?Element} */
+    this.rightLabel_ = null;
+
+    /** @private {!Element} */
```

Comment:
These should be set to null here and created in `buildCallback`. They type would be `{?Element}`. You can use `assertElement` later on to coerce types.